### PR TITLE
Package upgrade

### PR DESCRIPTION
**Pull request by kay0u**
## Solution
- *Fix all scripts*
- *Add change url*
- *Normalization from example_ynh*

## PR Status
- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Validation
---
- [ ] **Code review**
- [ ] **Approval (LGTM)**  
*Code review and approval have to be from a member of @YunoHost/apps group*
- **CI succeeded** : 
[Build Status/badge/icon)](https://ci-apps-dev.yunohost.org/jenkins/job/airsonic_ynh%20(Kay0u)/)

### CODE REVIEW

**Comment by kay0u** [comment]
This time, everything works well :-)

File: scripts/upgrade [hunk 1]
```diff
@@ -12,139 +12,155 @@ source /usr/share/yunohost/helpers
...
 
-### Verify the checksum of a file, stored by `ynh_store_file_checksum` in the install script.
-### And create a backup of this file if the checksum is different. So the file will be backed up if the admin had modified it.
+### For details, see https://airsonic.github.io/docs/transcode/
 
-# Path of the systemd service configuration
-ynh_backup_if_checksum_is_different "$final_path/airsonic.properties"
+mkdir -p $final_path/transcode
 
+if [ -x /usr/bin/ffmpeg -a ! -e $final_path/transcode/ffmpeg ]; then   # Check if 'ffmpeg' is installed and executable
```
Comment (Gofannon) [inline scripts/upgrade]: # Check if 'ffmpeg' is installed and executable and that symlink already exists

File: scripts/upgrade [hunk 2]
```diff
@@ -12,139 +12,155 @@ source /usr/share/yunohost/helpers
...
+	ln -s /usr/bin/ffmpeg $final_path/transcode
+fi
 
-# preload Airsonic configuration with LDAP settings
-# FIXME It does not work at the moment
-# Maybe the settings are wrong or this is an issue with the app itself.
-# Airsonic logs are not verbose enough even in TRACE mode
-# See https://airsonic.github.io/docs/logging/
+# FIXME Not sure if 'lame' is needed ?
+if [ -x /usr/bin/lame -a ! -e $final_path/transcode/lame ]; then   # Check if 'lame' is installed and executable
```
Comment (Gofannon) [inline scripts/upgrade]: # Check if 'lame' is installed and executable and that symlink already exists

File: scripts/change_url [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+#=================================================
+# SPECIFIC MODIFICATIONS
+#=================================================
+# SETUP SYSTEMD
+#=================================================
+ynh_script_progression --message="Configuring a systemd service..." --weight=3
+
+ynh_replace_string --match_string="__PORT__" --replace_string=$port --target_file="../conf/systemd.service"
+ynh_replace_string --match_string="__PATH__" --replace_string=$path_url --target_file="../conf/systemd.service"
```
Comment (Gofannon) [inline scripts/change_url]: You need to change variables in java command too https://github.com/YunoHost-Apps/airsonic_ynh/blob/package_upgrade/conf/systemd.service#L18

File: README.md [hunk 4]
```diff
@@ -1,69 +1,58 @@
...
-## Limitations
-
-* Any known limitations.
-
-## Additional information
-
-* Other information you would add about this application
+* x86-64b - [![Build Status](https://ci-apps.yunohost.org/ci/logs/airsonic%20%28Apps%29.svg)](https://ci-apps.yunohost.org/ci/apps/airsonic/)
+* ARMv8-A - [![Build Status](https://ci-apps-arm.yunohost.org/ci/logs/airsonic%20%28Apps%29.svg)](https://ci-apps-arm.yunohost.org/ci/apps/airsonic/)
+* Jessie x86-64b - [![Build Status](https://ci-stretch.nohost.me/ci/logs/airsonic%20%28Apps%29.svg)](https://ci-stretch.nohost.me/ci/apps/airsonic/)
```
[the remark this answers is not in the document]
Comment (kay0u) [reply README.md]: I just follow the template https://github.com/YunoHost/example_ynh

File: conf/systemd.service [hunk 5]
```diff
@@ -8,11 +8,11 @@ AssertPathExists=__FINALPATH__
...
 Environment="JAVA_OPTS=-Xmx700m"
-Environment="AIRSONIC_HOME=/var/airsonic"
-Environment="PORT=8080"
-Environment="CONTEXT_PATH=/airsonic"
+Environment="AIRSONIC_HOME=__FINALPATH__"
+Environment="PORT=__PORT__"
+Environment="CONTEXT_PATH=__PATH__"
 Environment="JAVA_ARGS="
 EnvironmentFile=-/etc/default/airsonic
 ExecStart=/usr/bin/java \
```
[the remark this answers is not in the document]
Comment (kay0u) [reply conf/systemd.service]: I don't know, I've never used this file, but I'm not really against its use.

I can restore the conf/systemd-sysconfig, and update scripts if you want

File: scripts/install [hunk 6]
```diff
@@ -159,135 +118,123 @@ fi
...
-
 yunohost service add $app --description "Airsonic daemon" --log "$final_path/airsonic.log"
 
+#=================================================
+# START SYSTEMD SERVICE
+#=================================================
+ynh_script_progression --message="Starting a systemd service..." --weight=12
+
+# Start a systemd service
+ynh_systemd_action --service_name=$app --action="start" --log_path="$final_path/airsonic.log" --line_match="Started Application in"
```
[the remark this answers is not in the document]
Comment (kay0u) [reply scripts/install]: This line is to pass package_check, and be sure that airsonic started at the end of the install.

The thing you have to know to use this feature is that you have to add this line:
https://github.com/YunoHost-Apps/airsonic_ynh/pull/1/files/a8fcc3f55cbc273ef1b08832772d00f8c79ae92d#diff-ad962c6fd65b725b71aa46c8a715127fR17

Doc here:
https://github.com/YunoHost/yunohost/blob/stretch-unstable/data/helpers.d/systemd#L79-L81

File: scripts/install [hunk 7]
```diff
@@ -159,135 +118,123 @@ fi
...
+
+#http://www.subsonic.org/pages/api.jsp
+passwordAdmin="admin"
+salt=$(ynh_string_random --length=8)
+token=$(echo -n $passwordAdmin$salt | md5sum | awk '{print $1}')
+
+# Installation with curl
+ynh_script_progression --message="Finalizing installation..." --weight=6
+
+VERSION="1.15.0"
```
[the remark this answers is not in the document]
Comment (kay0u) [reply scripts/install]: You're right, I must add doc here!

I found this information at this location:

https://airsonic.github.io/docs/api/

> Airsonic currently embed Subsonic API version 1.15.0.

And in the Subsonic API doc:

http://www.subsonic.org/pages/api.jsp

>  Starting with API version 1.13.0, the recommended authentication scheme is to send an authentication token, calculated as a one-way salted hash of the password.
This involves two steps:
> 1. For each REST call, generate a random string called the salt. Send this as parameter `s`. Use a salt length of at least six characters.
> 2. Calculate the authentication token as follows: token = md5(password + salt). The md5() function takes a string and returns the 32-byte ASCII hexadecimal representation of the MD5 hash, using lower case characters for the hex values. The '+' operator represents concatenation of the two strings. Treat the strings as UTF-8 encoded when calculating the hash. Send the result as parameter `t`.

>For example: if the password is sesame and the random salt is c19b2d, then token = md5("sesamec19b2d") = 26719a1196d2a940705a59634eb18eab. The corresponding request URL then becomes:
http://your-server/rest/ping.view?u=joe&t=26719a1196d2a940705a59634eb18eab&s=c19b2d&v=1.12.0&c=myapp

File: scripts/change_url (hunk 3 above)
Comment (kay0u) [reply scripts/change_url]: No, I don't think so..... Or maybe I didn't understand what you meant. Can you be more precise please?

**Comment by kay0u** [comment]
> Instead of postgres, why not using mysql / mariadb ?

With the current version packaged, postgres/mysql works well, but this the new one, it crashes at startup for no reason. I think there is a bug in airsonic. This feature could be added later, but it is not really necessary (or could be for performance reasons).

File: conf/systemd.service (hunk 5 above)
[the remark this answers is not in the document]
Comment (kay0u) [reply conf/systemd.service]: I reuse the file `conf/systemd-sysconfig`

**Comment by kay0u** [comment]
Package check is running, ready for review :-)

**Comment by kay0u** [comment]
Everything is working with package_check (I don't know why https://ci-apps-dev.yunohost.org explode, but it's not due to airsonic I think)

**Comment by Gofannon** [comment]
Seems good, thanks.
I guess that I can merge the branch directly. Tell me if I'm wrong :)

I will try to find time to try the "testing version"

**Comment by kay0u** [comment]
You're welcome!

Ping me if you find an issue or need some help! :-)

Don't forget to create a PR to set the status of airsonic as "working" [apps.json](https://github.com/YunoHost/apps/blob/f5efb33d717219ae0ccb8e0bf47a390a5eef0afa/apps.json#L68)  once the testing branch is merged.

**Comment by Gofannon** [comment]
Alright, will do!

**Pull request merged**